openmeter/billing/adapter/invoicelines.go

```diff
@@ -0,0 +1,119 @@
+package billingadapter
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openmeterio/openmeter/openmeter/billing"
+	billingentity "github.com/openmeterio/openmeter/openmeter/billing/entity"
+	"github.com/openmeterio/openmeter/openmeter/ent/db"
+	"github.com/openmeterio/openmeter/openmeter/ent/db/billinginvoiceline"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+var _ billing.InvoiceLineAdapter = (*adapter)(nil)
+
+func (r *adapter) CreateInvoiceLines(ctx context.Context, input billing.CreateInvoiceLinesAdapterInput) (*billing.CreateInvoiceLinesResponse, error) {
+	result := &billing.CreateInvoiceLinesResponse{
+		Lines: make([]billingentity.Line, 0, len(input.Lines)),
+	}
+
+	for _, line := range input.Lines {
+		newEnt := r.db.BillingInvoiceLine.Create().
+			SetNamespace(input.Namespace).
+			SetInvoiceID(line.InvoiceID).
+			SetPeriodStart(line.Period.Start).
+			SetPeriodEnd(line.Period.End).
+			SetInvoiceAt(line.InvoiceAt).
+			SetStatus(line.Status).
+			SetType(line.Type).
+			SetName(line.Name).
+			SetCurrency(line.Currency).
+			SetTaxOverrides(line.TaxOverrides).
+			SetNillableQuantity(line.Quantity).
+			SetMetadata(line.Metadata)
+
+		edges := db.BillingInvoiceLineEdges{}
+
+		switch line.Type {
+		case billingentity.InvoiceLineTypeManualFee:
+			// Let's create the manual line for the invoice
+			newManualLineConfig, err := r.db.BillingInvoiceManualLineConfig.Create().
+				SetNamespace(input.Namespace).
+				SetUnitPrice(line.ManualFee.Price).
+				Save(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			newEnt = newEnt.SetBillingInvoiceManualLines(newManualLineConfig)
+			edges.BillingInvoiceManualLines = newManualLineConfig
+		default:
+			return nil, fmt.Errorf("unsupported type: %s", line.Type)
+		}
+
+		savedLine, err := newEnt.Save(ctx)
```

Comment:
```
var linesToCreate ent.BillingInvoiceLine[]
for _, line := range input.Lines {
   linesToCreate = append(linesToCreate, r.db.BillingInvoiceLine.Create()
      ...
    )
}

savesLines, err := newEnt.CreateBulk(...linesToCreate)
```
https://entgo.io/docs/crud/#create-many

Reply:
I have created a ticket on this, but given that I have to create a lot of additional objects, I wouldn't do that refactor just now. (We are not creating 10000s of records).